clab/general.go

```diff
@@ -0,0 +1,126 @@
+package clab
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/cloudflare/cfssl/log"
+	"github.com/srl-labs/containerlab/types"
+)
+
+var (
+	CLAB_HOSTENTRY_PREFIX  = "###### CLAB-%s-START ######"
+	CLAB_HOSTENTRY_POSTFIX = "###### CLAB-%s-END ######"
+)
+
+func AppendHostsFileEntries(containers []types.GenericContainer, labname string) error {
+	if labname == "" {
+		return fmt.Errorf("missing lab name")
+	}
+	// lets make sure we do not have remaining of a non destroyed run in the hosts file
+	err := DeleteEntriesFromHostsFile(labname)
+	if err != nil {
+		return err
+	}
+	data := GenerateHostsEntries(containers, labname)
+	if len(data) == 0 {
+		return nil
+	}
+	f, err := os.OpenFile("/etc/hosts", os.O_APPEND|os.O_WRONLY, os.ModeAppend)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = f.WriteString("\n")
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// hostEntries builds an /etc/hosts compliant text blob (as []byte]) for containers ipv4/6 address<->name pairs
+func GenerateHostsEntries(containers []types.GenericContainer, labname string) []byte {
+	entries := strings.Builder{}
+	v6entries := strings.Builder{}
+
+	entries.WriteString(fmt.Sprintf(CLAB_HOSTENTRY_PREFIX+"\n", labname))
+
+	for _, cont := range containers {
+		if len(cont.Names) == 0 {
+			continue
+		}
+		if cont.NetworkSettings.Set {
+			if cont.NetworkSettings.IPv4addr != "" {
+				fmt.Fprintf(&entries, "%s\t%s\n", cont.NetworkSettings.IPv4addr, strings.TrimLeft(cont.Names[0], "/"))
+			}
+			if cont.NetworkSettings.IPv6addr != "" {
+				fmt.Fprintf(&v6entries, "%s\t%s\n", cont.NetworkSettings.IPv6addr, strings.TrimLeft(cont.Names[0], "/"))
+			}
+		}
+	}
+
+	entries.WriteString(v6entries.String())
+	entries.WriteString(fmt.Sprintf(CLAB_HOSTENTRY_POSTFIX+"\n", labname))
+	return []byte(entries.String())
+}
+
+func DeleteEntriesFromHostsFile(labname string) error {
+	if labname == "" {
+		return fmt.Errorf("missing containerlab name")
+	}
+	f, err := os.OpenFile("/etc/hosts", os.O_RDWR, 0644) // skipcq: GSC-G302
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Info("/etc/hosts does not exist")
```

Comment:
?
```suggestion
			log.Info("/etc/hosts does not exist")
			return nil
```

Reply:
@hellt how do you want it to be?

Reply:
if /etc/hosts doesn't exist that looks like an error to me as far as this function is concerned

@kellerza why do you want to return `nil` in that case and not `errors.New("/etc/hosts does not exist")`?

Reply:
If you return an error then the calling function (AppenHostEntries) will fail to append.

So we either return a nil here, or we change AppenHostentries to still write if the file does not exist

Reply:
if there is no hosts file, where the `AppendHostEntries` will append to?

Reply:
I would assume it should create the hosts file?

Reply:
It does not so far. Is creating the file what we want to happen?